src/classes/query.py

```diff
@@ -157,6 +166,10 @@ def save(self):
         """ Save the object back to the project data store """
         super().save(Clip)
 
+    def genid(self):
+        """ Generate new id for the object """
+        super().genid(Clip)
+
```

Comment:
So it's an important part of batch inserting. Do you think there is another way?

Reply:
@rabits Well, I think the way the current code does it is probably right, as lame as it seems. It just blanks the `id` key of the copied dictionary, which then allows a new one to be generated on `.save()`. There's a lot of special handling in `classes.updates` (see `get_reverse_action`, primarily) for "insert" and/or "delete" type actions, including ID management. `old_values` gets swapped into `values` for the transaction when it's reversed, which is why it needs to be managed properly.

But I don't think avoiding `Clip.save()` to talk directly to the update manager is the way to do this. The update manager needs the concept of a _transaction_ that combines **multiple** `UpdateAction`s (rather than an `UpdateAction` that contains multiple values) — a while ago [I sketched out](https://github.com/OpenShot/openshot-qt/pull/3178#issuecomment-575375343) how that might work, inspired largely by Qt's own `QUndoStack`.

The key thing was, by adopting an _implicit_ transaction functionality (triggered by explicit calls to methods that would begin and end a transaction), things like `QueryObject.save()` could still be used... unchanged, even, for the most part. There'd be no reason to avoid using it just to batch up multiple `.save()`s.

If `Clip.save()` (or `Transition.save()` or anything else) is called between `get_app().updates.begin_transaction()` and `get_app().updates.end_transaction()`, its insert/update actions would be redirected onto a transaction stack, instead of the main history (undo) stack. The `end_transaction()` call would then do the work of creating the actual history entry containing all of the batched items. The Undo and Redo processing would also have to understand how to process those batched items.